### PR TITLE
Optimize string.GetFileFromFilename

### PR DESCRIPTION
**Pull request by afxnatic**
no longer uses a loop and has far fewer C calls as a result

### CODE REVIEW

File: garrysmod/lua/includes/extensions/string.lua [hunk 1]
```diff
@@ -164,12 +164,19 @@ end
 	Usage: string.GetFileFromFilename("garrysmod/lua/modules/string.lua")
 -----------------------------------------------------------]]
 function string.GetFileFromFilename( path )
-	for i = #path, 1, -1 do
-		local c = string.sub( path, i, i )
-		if ( c == "/" or c == "\\" ) then return string.sub( path, i + 1 ) end
+
+	local reversePath = string.reverse( path )
+
+	-- Way faster than fetching length and calling two C functions per character.
```
Comment (robotboy655) [inline garrysmod/lua/includes/extensions/string.lua:170]: "meta" comments like this should not be included in code.

File: garrysmod/lua/includes/extensions/string.lua [hunk 2]
```diff
@@ -164,12 +164,19 @@ end
...
 -----------------------------------------------------------]]
 function string.GetFileFromFilename( path )
-	for i = #path, 1, -1 do
-		local c = string.sub( path, i, i )
-		if ( c == "/" or c == "\\" ) then return string.sub( path, i + 1 ) end
+
+	local reversePath = string.reverse( path )
+
+	-- Way faster than fetching length and calling two C functions per character.
+	local slashPosition, _, _ = string.find( reversePath, "/", 1 ) or string.find( reversePath, "\\", 1 )
```
Comment (robotboy655) [inline garrysmod/lua/includes/extensions/string.lua:171]: Adding `true` for the 3rd param should make this even faster.

**Review by robotboy655 — CHANGES_REQUESTED** [review]
Your implementation with the following input does not produce desired output:
```
print( "OLD", string.GetFileFromFilename( "garrysmod/lua/modules\\string.lua" ) )
print( "NEW", string.GetFileFromFilename2( "garrysmod/lua/modules\\string.lua" ) )

OLD	string.lua
NEW	modules\string.lua
```

**Comment by afxnatic** [comment]
> Your implementation with the following input does not produce desired output:
> 
> ```
> print( "OLD", string.GetFileFromFilename( "garrysmod/lua/modules\\string.lua" ) )
> print( "NEW", string.GetFileFromFilename2( "garrysmod/lua/modules\\string.lua" ) )
> 
> OLD	string.lua
> NEW	modules\string.lua
> ```

test with 222193e8d288b14a70b808061161e1308ddf7f95 and this issue should no longer be present

**Comment by robotboy655** [comment]
You cannot just flip the string,find's and expect it to work:
```
print( "OLD", string.GetFileFromFilename( "garrysmod/lua\\modules/string.lua" ) )
print( "OLD", string.GetFileFromFilename( "garrysmod/lua/modules\\string.lua" ) )

print( "NEW", string.GetFileFromFilename2( "garrysmod/lua\\modules/string.lua" ) )
print( "NEW", string.GetFileFromFilename2( "garrysmod/lua/modules\\string.lua" ) )
```

```
OLD	string.lua
OLD	string.lua
NEW	modules/string.lua
NEW	string.lua
```

**Comment by afxnatic** [comment]
> You cannot just flip the string,find's and expect it to work:

string.find is actually flawed for any path that mixes backslashes and forward slashes, my bad for not realizing that

i replaced the existing string.find solution with a lua pattern in 886ed13251a46be2b026c55531dc2eed6ff07db1
```
print( "OLD", string.GetFileFromFilename( "garrysmod/lua/modules/string.lua" ) )
print( "OLD", string.GetFileFromFilename( "garrysmod/lua\\modules/string.lua" ) )
print( "OLD", string.GetFileFromFilename( "garrysmod/lua/modules\\string.lua" ) )

print( "NEW", string.GetFileFromFilename2( "garrysmod/lua/modules/string.lua" ) )
print( "NEW", string.GetFileFromFilename2( "garrysmod/lua\\modules/string.lua" ) )
print( "NEW", string.GetFileFromFilename2( "garrysmod/lua/modules\\string.lua" ) )
```

```
OLD	string.lua
OLD	string.lua
OLD	string.lua
NEW	string.lua
NEW	string.lua
NEW	string.lua
```

it even keeps this issue:
```
print( "OLD", string.GetFileFromFilename( "garrysmod/lua/modules" ) )
print( "NEW", string.GetFileFromFilename2( "garrysmod/lua/modules" ) )
```

```
OLD	modules
NEW	modules
```

if its still buggy or doesn't match original behavior, just close this PR

**Comment by robotboy655** [comment]
Using Lua patterns is significantly slower than the current code. It's why these functions were changed to not use them in the first place. (https://github.com/Facepunch/garrysmod/commit/7017e076718a953aa9bfdc9ee4f3b490d6852cea)

image
Your current proposal also has a regression when the string has no slashes.

Using 2 string.find calls can be faster in some cases, but also slower in others.

**Pull request closed without merging**